Add security documentation (#105)
canonical/hockeypuck-k8s-operator: docs/explanation/security.md

# Security

This document outlines common risks and possible best practices for the Hockeypuck charm. It focuses on configurations and protections available through the charm itself.

For details regarding upstream Hockeypuck configuration and broader security considerations, please refer to the [official Hockeypuck documentation](https://hockeypuck.io/configuration.html).

## Risks

The following items include descriptions of the risks, their corresponding best practices for mitigation, as well as links to related documentation and configuration guidelines.

### Loss of data

> Hockeypuck database might be destroyed, corrupted, or become inaccessible.

Reviewer: ```suggestion
The Hockeypuck database might become inaccessible, corrupted, or may be destroyed.
```
I think that "become" also works for "corrupted". I also rearranged the order in terms of (my perceived) severity -- becoming inaccessible is not great, but being destroyed is worse.

#### Best practices

- Set up regular backups:

  Follow the [charm documentation](https://charmhub.io/hockeypuck-k8s/docs/how-to-backup-and-restore-hockeypuck) for guidance on creating regular backups and restoring them when required.
  

>   Hockeypuck charm maintains a table `deleted_keys` that contains a list of all the blocklisted keys. This table is read and fed into the `hockypuck.conf` file used by Hockeypuck during the startup process. If this data is lost, Hockeypuck will not block any keys during startup. It is important to have a backup of the charm to avoid the blocked keys from being reconciled while peering.

Reviewer: ```suggestion
  The Hockeypuck charm maintains a table `deleted_keys` that contains a list of all the blocklisted keys. This table is read and fed into the `hockypuck.conf` file used by Hockeypuck during the startup process. If this data is lost, Hockeypuck will not block any keys during startup. It is important to have a backup of the charm to avoid the blocked keys from being reconciled while peering.
```

- Avoid manual key deletion:

>   Manually deleting PGP keys from the PostgreSQL database does not remove the key entirely from Hockeypuck. Hockeypuck maintains another leveldb database locally apart from the PostgreSQL database to build the prefix tree. Deleting the key from PostgreSQL does not delete the key from the leveldb database. To ensure the key is properly removed from Hockeypuck, refer to [Hockeypuck's CharmHub documentation](https://charmhub.io/hockeypuck-k8s/docs/how-to-manage-gpg-keys) on how to use Hockeypuck's `delete` API or the `block-keys` action.

Reviewer: ```suggestion
  Manually deleting PGP keys from the PostgreSQL database does not remove the key entirely from Hockeypuck. Hockeypuck maintains another leveldb database locally apart from the PostgreSQL database to build the prefix tree. Deleting the key from PostgreSQL does not delete the key from the leveldb database. To ensure the key is properly removed from Hockeypuck, refer to [How to manage GPG keys](https://charmhub.io/hockeypuck-k8s/docs/how-to-manage-gpg-keys) on how to use Hockeypuck's `delete` API or the `block-keys` action.
```
I guess that https://charmhub.io/hockeypuck-k8s/docs/how-to-manage-gpg-keys is a stand-in link while we wait for the docs to be available on Charmhub?

Author: yes that is correct.

### Denial-of-Service (DoS) attacks

> A denial-of-service attack could overwhelm Hockeypuck with traffic, preventing legitimate users from accessing the service. Following the recent [DoS attack on keyservers](https://gist.github.com/rjhansen/67ab921ffb4084c865b3618d6955275f), implementing the [DOS protection rules](https://github.com/hockeypuck/hockeypuck/tree/master/contrib/docker-compose/standalone/haproxy/etc) provided by Hockeypuck is mandatory to allow peering with other external servers.

Reviewer: ```suggestion
A denial-of-service attack could overwhelm Hockeypuck with traffic, preventing legitimate users from accessing the service. Following the recent [DoS attack on keyservers](https://gist.github.com/rjhansen/67ab921ffb4084c865b3618d6955275f), implementing the [DoS protection rules](https://github.com/hockeypuck/hockeypuck/tree/master/contrib/docker-compose/standalone/haproxy/etc) provided by Hockeypuck is mandatory to allow peering with other external servers.
```

[note]

> The current Hockeypuck charm does not have the DOS protection rules implemented due to implementation constraints and hence will not be able to reconcile with external peers. The DOS protection will be incorporated in future releases.

Reviewer: ```suggestion
The current Hockeypuck charm does not have the DoS protection rules implemented due to implementation constraints and hence will not be able to reconcile with external peers. The DoS protection will be incorporated in future releases.
```

[/note]

### Security vulnerabilities

> Running Hockeypuck with one or more weakness that can be exploited by attackers.

Reviewer: ```suggestion
Running Hockeypuck with one or more weaknesses can be exploited by attackers.
```
Should we give examples or be more explicit about what these weaknesses are?
